orttraining/tools/scripts/pipeline_model_split.py

```diff
@@ -0,0 +1,376 @@
+
+    input_wait.input.extend(input_tensors)
+    input_wait.output.extend([i + '_sync' for i in input_tensors])
+
+    # split the graph based on edgeIds
+    upstream_nodes = []
+    output_shapes =[]
+
+    for id in edgeIds:
+        for node in model.graph.node:
+            if len(node.output) >=1 and node.output[0] == id:
+                upstream_nodes.append(node)
+                for info in model.graph.value_info:
+                    if info.name == id:
+                        output_shapes.append(info.type)
+
+    record_signal = model.graph.input.add()
+    record_signal.CopyFrom(helper.make_tensor_value_info(
+        'record_signal', onnx.TensorProto.INT64, None))
+
+    wait_signal = model.graph.input.add()
+    wait_signal.CopyFrom(helper.make_tensor_value_info(
+        'wait_signal', onnx.TensorProto.INT64, None))
+
+    send_signal = model.graph.input.add()
+    send_signal.CopyFrom(helper.make_tensor_value_info(
+        'send_signal', onnx.TensorProto.BOOL, None))
+
+    recv_signal = model.graph.input.add()
+    recv_signal.CopyFrom(helper.make_tensor_value_info(
+        'recv_signal', onnx.TensorProto.BOOL, None))
+
+    # output signal from send after cut
+    send_output_signal = model.graph.output.add()
+    send_output_signal.CopyFrom(helper.make_tensor_value_info(
+        'send_output_signal', onnx.TensorProto.BOOL, None))
+
+    # output signal from receive after cut
+    receive_output_signal = model.graph.output.add()
+    receive_output_signal.CopyFrom(helper.make_tensor_value_info(
+        'receive_output_signal', onnx.TensorProto.BOOL, None))
+
+    new_send = model.graph.node.add()
+    new_send.CopyFrom(helper.make_node(
+        'Send',
+        inputs=['send_signal'],
+        outputs=['send_output_signal'],
+        tag=0,
+        src=0, # todo：change it when split more than 2 parts
+        dst=1,
+        domain=ms_domain,
+        element_type=1, # assuming all tensors are of type float
+        name='send'))
+
+    new_receive = model.graph.node.add()
+    new_receive.CopyFrom(helper.make_node(
+        'Recv',
+        inputs=['recv_signal'],
+        outputs=['receive_output_signal'],
+        tag=1,
+        src=0, # todo：change it when split more than 2 parts
+        dst=1,
+        domain=ms_domain,
+        element_type=1, # assuming all tensors are of type float
+        name='receive'))
+
+    new_wait = model.graph.node.add()
+    new_wait.CopyFrom(helper.make_node(
+        'WaitEvent',
+        inputs=['wait_signal'],
+        outputs=[],
+        domain=ms_domain))
+
+    new_record = model.graph.node.add()
+    new_record.CopyFrom(helper.make_node(
+        'RecordEvent',
+        inputs=['record_signal'],
+        outputs=[],
+        domain=ms_domain))
+
+    for i in range(len(upstream_nodes)):
+        n = upstream_nodes[i]
+        output_type = output_shapes[i]
+
+        output_nodes = find_all_output_nodes_by_edge(model, n.output[0])
```

Comment:
>n.output[0] [](start = 60, length = 11)

is this always output[0]?

Reply:
I am making the assumption that each op has only 1 output, so ouput[0] is all the outputs this op has. Surely it won't work with graph with ops like TopK etc. 
But it seems other scipt in this folder is also making such assumption, so not sure how often such case would appear in transformer model. 

---
In reply to: [401059154](https://github.com/microsoft/onnxruntime/pull/3275#discussion_r401059154) [](ancestors = 401059154)

Reply:
Even for BERT, we have have nodes like LayerNormalization that have multiple outputs. If we want the cut to be generic, the assumption of output[0] no longer exists.

---
In reply to: [401166157](https://github.com/microsoft/onnxruntime/pull/3275#discussion_r401166157) [](ancestors = 401166157,401059154)